Snap updates geneva

tonyespy opened the pull request:
## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [x] Other... Please describe:

This PR includes a number of commits which update the snap packaging for Geneva. Some of these are bug fixes recently made to the fuji branch, some address updating configuration files for new Geneva behavior (e.g. security-secretstore-setup's new token provider for per-service vault tokens), and some are general cleanup of the snap packaging.

## Does this PR introduce a breaking change?
<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

- [ ] Yes
- [x] No

## Other information

anonymouse64 commented:
Unclear how this happened, but snapcraft wasn't able to checkout the specific commit in snapcraft-preload:

```
01:17:51 Pulling snapcraft-preload 
01:17:53 Cloning into '/build/parts/snapcraft-preload/src'...
01:17:54 fatal: reference is not a tree: d654bbe8a5add5a4bea14d96342b656990c5c818
01:17:54 Failed to pull source, command 'git -C /build/parts/snapcraft-preload/src checkout d654bbe8a5add5a4bea14d96342b656990c5c818' exited with code 128.
01:17:59 Build step 'Execute shell' marked build as failure
```

I couldn't reproduce this locally, so the tests here probably just need to be re-run

anonymouse64 commented:
Hmm actually I can reproduce this with snapcraft stable, I was not on snapcraft stable...

tonyespy commented:
> Look good but I suspect there are some initialization race conditions that haven't been dealt with. In particular, if you enable security-file-token-provider, the first first thing it does when it comes up is revoke all the tokens it issued the last time it was launched. This means there needs to be some kind of init blocking on the other services such they won't start until security-secretstore-setup is done. This may have been done in the previous release and so it isn't showing up in the diff, but I would feel more confortable about the changes if you could point me to the logic that shows that this handshaking is coded in somewhere because it is still not running perfectly in even the docker configuration.

Thanks for the review @bnevis-i In my opinion it is not necessary to adopt the per-service tokens as there is no software layer the prevent one microservice from reading another microservice's secrets-token.json since they are all part of the same snap.! As for the service ordering, snapcraft supports the use of **After**/**Before** keywords for services, which are then used in the dynamically generated systemd service units created for every service in the snap.

So if you look at the service definitions for edgex-mongo, and security-proxy-setup, you'll notice they're gated with:

after:
  - security-secretstore-setup
  - kong-daemon

You'll also notice in turn, that most of the other services are gated on security-proxy-setup, edgex-mongo, and core-config-seed. So, unless I'm missing something, I think we're all set with regards to token race conditions and the snap.

> In my opinion it is not necessary to adopt the per-service tokens as there is no software layer the prevent one microservice from reading another microservice's secrets-token.json since they are all part of the same snap.


With regards to whether it's even appropriate to require the services in the same snap to use different tokens, fair point. I'm fine reverting back to a model where a single shared token is used for the services within the snap, however I'll need some help with how to specify that via the relevant configuration files. Right now, by default, all the services (including proxy-setup, and edgex-mongo) are configured by default for per-service tokens, so on install, we end up with service failure(s) (actually just one, as the first failed service causes the install/update to fail, and auto-revert to the previous version).

One other thing that will need to be addressed is if/when we have additional snaps installed (e.g. application or device services)? In this case, we'll need to add a content interface that will allow the token provider to export per-snap tokens to other snaps. That said, we have some time to work out that use case.

bnevis-i commented:
> As for the service ordering, snapcraft supports the use of After/Before keywords for services....  So, unless I'm missing something, I think we're all set with regards to token race conditions and the snap.

Great news.

> Right now, by default, all the services (including proxy-setup, and edgex-mongo) are configured by default for per-service tokens, so on install, we end up with service failure(s)

Are you saying the PR as it stands yields in a non-working configuration?  Feel free to reach out directly to me if this is the case as it sounds a little too complex to debug in the PR.

>  In this case, we'll need to add a content interface that will allow the token provider to export per-snap tokens to other snaps. That said, we have some time to work out that use case.

Per-service tokens are very amenable to this, so I am supportive of staying the course and work on any issues that might crop up.

bnevis-i commented on snap/snapcraft.yaml:
```diff
@@ -795,23 +789,43 @@ parts:
+            -e "s@CaFilePath = '/tmp/edgex/secrets/ca/ca.pem'@CaFilePath = \"\$SNAP_DATA/pki/ca/ca.pem\"@" \
+            -e "s@PrivilegedTokenPath = \"/run/edgex/secrets/tokenprovider/secrets-token.json\"@PrivilegedTokenPath = \"\$SNAP_DATA/secrets/tokenprovider/secrets-token.json\"@" \
+            -e "s@ConfigFile = \"res-file-token-provider/token-config.json\"@ConfigFile = \"\$SNAP_DATA/config/security-file-token-provider/res/token-config.json\"@" \
+            -e "s@OutputDir = \"/tmp/edgex/secrets\"@OutputDir = \"\$SNAP_DATA/secrets\"@" \
```
OutputDir is the root directory  of the per-service secrets

There is a subdirectory for each service named after the service key (e.g. edgex-core-data).
Inside of that folder is that service's bootstrapping secrets.  `file-token-provider` creates each service's `secrets-token.json` there.  security-secrets-setup also creates PKI assets (such as the ca certificate and leaf certs for kong and vault) under there using their respective service keys.

bnevis-i replied:
See https://github.com/edgexfoundry/edgex-docs/blob/master/docs_src/microservices/security/security-file-token-provider.1.md

anonymouse64 replied:
Thanks Bryon, this matches the rest of the filepaths in the config file we're setting up here

bnevis-i commented:
> This LGTM, but doesn't install properly:
> 
> ```
05T16:56:37.028387006Z app=edgex-mongo source=main.go:57 msg="edgex-mongo failed because of: open /vault/config/assets/resp-init.json: no such file or directory"
> ```

Definitely missed a rewrite here.

tonyespy commented:
> > As for the service ordering, snapcraft supports the use of After/Before keywords for services....  So, unless I'm missing something, I think we're all set with regards to token race conditions and the snap.
> 
> Great news.
> 
> > Right now, by default, all the services (including proxy-setup, and edgex-mongo) are configured by default for per-service tokens, so on install, we end up with service failure(s)
> 
> Are you saying the PR as it stands yields in a non-working configuration? Feel free to reach out directly to me if this is the case as it sounds a little too complex to debug in the PR.

No, the PR does work, as all the services are now starting properly, whereas prior to this PR, the snap wouldn't install due to services failing on startup due to not being able to read their token files. So given that's the case, I'd prefer we landed this PR as is, and then if we still think it makes sense to go back to a single token for all, I can work with you to update the configuration files accordingly.
> 
> > In this case, we'll need to add a content interface that will allow the token provider to export per-snap tokens to other snaps. That said, we have some time to work out that use case.
> 
> Per-service tokens are very amenable to this, so I am supportive of staying the course and work on any issues that might crop up.

tonyespy commented:
@anonymouse64 I've re-pushed with the removal of "source-depth" as we discussed earlier. I also fixed the issue with edgex-mongo not starting. It turns out we made yet another breaking change to a configuration file due to this issue:

https://github.com/edgexfoundry/docker-edgex-mongo/issues/68

I've updated the snap's sed script which updates the CaCert and TokenFile/Path to point at directories under $SNAP_DATA, so all appears to be working once again.

tonyespy commented:
I also pushed one more commit to handle the recent default db change.

anonymouse64 commented on snap/hooks/configure:
```diff
@@ -70,7 +68,7 @@ for key in $ALL_SERVICES; do
             handle_svc "$key" "$status"
             ;;
-        mongo-worker|security-proxy-setup|security-secretstore-setup)
+        security-proxy-setup|security-secretstore-setup)
```
good catch on this

anonymouse64 commented on snap/hooks/install:
```diff
@@ -189,7 +189,7 @@ snapctl stop "$SNAP_NAME.postgres"
 # off.
```
the comment in the line above this is wrong now

tonyespy replied:
Just changed the comment about disabling services:

```
 # finally, disable and turn off non-default services
-# by default, we want the support-*, device-*, and redis services
+# by default, we want the support-*, device-*, and mongo services

```
